calyx/opt/src/passes_experimental/fsm_builder.rs

```diff
@@ -0,0 +1,276 @@
+use crate::analysis::{IncompleteTransition, StaticSchedule};
+use crate::traversal::{Action, ConstructVisitor, Named, Visitor};
+use calyx_ir::{self as ir, GetAttributes};
+use calyx_utils::CalyxResult;
+const ACYCLIC: ir::Attribute =
+    ir::Attribute::Internal(ir::InternalAttr::ACYCLIC);
+
+pub struct FSMBuilder {}
+
+impl Named for FSMBuilder {
+    fn name() -> &'static str {
+        "fsm-builder"
+    }
+    fn description() -> &'static str {
+        "generates medium fsms in one pass for static and dynamic"
+    }
+}
+
+impl ConstructVisitor for FSMBuilder {
+    fn from(_ctx: &ir::Context) -> CalyxResult<Self> {
+        Ok(FSMBuilder {})
+    }
+    fn clear_data(&mut self) {}
+}
+
+// node that passes into this function is a part of the schedule
+impl StaticSchedule<'_, '_> {
+    fn build_abstract(
+        &mut self,
+        scon: &ir::StaticControl,
+        guard: ir::Guard<ir::Nothing>,
+        mut transitions_to_curr: Vec<IncompleteTransition>,
+        looped_once_guard: Option<ir::Guard<ir::Nothing>>,
+    ) -> (Vec<IncompleteTransition>, Option<ir::Guard<ir::Nothing>>) {
+        match scon {
+            ir::StaticControl::Empty(_) => (transitions_to_curr, None),
+            ir::StaticControl::Enable(sen) => {
+                if matches!(sen.get_attributes().get(ACYCLIC), Some(1)) {
+                    // allocate one state per cycle
+                    // for all parts of the FSM that want to transition to this enable,
+                    // register their transitions in self.state2trans
```

Comment:
I'm a little confused by this comment: `self.state2trans` comes from another file, and it's not mentioned anywhere else in this file. Maybe this could explain in a little more detail about where the data is going?